### PR TITLE
Attempt to fix error caused by same sprite found in two spritesheets

### PR DESCRIPTION
**Pull request by marekventur**
```
Error: Sprite word_option_button_00018 exists twice in manifest web_retina. Contained groups: game, game
```

My current theory as to why this is happening is that due to some file system shenigans the order of sprites processed has changed. Since the group hash is based on a sorted list of sprites this will mean the group hash stays the same. It might however change which sprites gets put into which spritesheet, potentially leading to a case where only some spritesheets get reprossesed and certain sprites now being found in two spritesheets.

This fix adds the whole set of sprite paths to the spritesheet hash invalidating them when the collection of sprites on that sheet changes.

Disclaimer: I have so far not been able to replicate this locally, so this fix is an educated guess.

This will change the hash key for all spritesheets, so expect longer build times for the one build after the update.

### CODE REVIEW

File: lib/spritesheet.js [hunk 1]
```diff
@@ -83,11 +84,17 @@ module.exports = function Spritesheet(name, scaledSprites, groupConfig, config,
         that.outputFilesize = data.outputFilesize;
     }
 
+     function calculateHash() {
+        var hash = crypto.createHash("sha1");
+        hash.update(_.pluck(scaledSprites, "path").sort().join(" ") + "_");
```
Comment (marekventur) [inline lib/spritesheet.js]: This is the new bit added to the hash.

File: lib/spritesheet.js [hunk 2]
```diff
@@ -83,11 +84,17 @@ module.exports = function Spritesheet(name, scaledSprites, groupConfig, config,
...
+        hash.update(that.extension + "_");
+        hash.update((groupConfig.quality ? groupConfig.quality : "noquality") + "_");
+        hash.update((groupConfig.compression ? groupConfig.compression : "nocompression"));
```
Comment (uorbe001) [inline lib/spritesheet.js]: doesn't update override the data on the hash? wouldn't this mean that the only 'update' that is actually relevant? (the rest would be overriden by the last one)

File: lib/spritesheet.js [hunk 3]
```diff
@@ -83,11 +84,17 @@ module.exports = function Spritesheet(name, scaledSprites, groupConfig, config,
...
+        return name + "_" + hash.digest("hex");
+    }
+
```
Comment (uorbe001) [inline lib/spritesheet.js]: could you add a unit test for this? it's quite hard to understand this code without understanding what the input values are, and that's not easy to find

Comment (marekventur) [inline lib/spritesheet.js]: yes, that's a good idea.

File: lib/spritesheet.js (hunk 2 above)
Comment (marekventur) [inline lib/spritesheet.js]: No: https://nodejs.org/api/crypto.html#crypto_hash_update_data_input_encoding

**Comment by marekventur** [comment]
@uorbe001 are you happy with this? I could do a shared release together with #4, it would save me the work of having to update all the games twice.

File: test/unit/full.js [hunk 4]
```diff
@@ -37,12 +37,12 @@ describe("Full run", function () {
...
     });
 
-    after(function() {
+    /*after(function() {
         return Q.nfcall(rimraf, outputPath)
         .then(function() {
             return Q.nfcall(rimraf, tempPath);
         });
-    });
+    });*/
```
Comment (uorbe001) [inline test/unit/full.js]: why is this commented out?

**Comment by uorbe001** [comment]
@marekventur just one comment

File: test/unit/full.js (hunk 4 above)
Comment (marekventur) [inline test/unit/full.js]: debugging. good catch.

**Pull request merged**